src/workerd/api/encoding.c++

```diff
@@ -464,25 +465,74 @@ jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
   return js.alloc<TextEncoder>();
 }
 
-namespace {
-TextEncoder::EncodeIntoResult encodeIntoImpl(
-    jsg::Lock& js, jsg::JsString input, jsg::BufferSource& buffer) {
-  auto result = input.writeInto(
-      js, buffer.asArrayPtr().asChars(), jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
-  return TextEncoder::EncodeIntoResult{
-    .read = static_cast<int>(result.read),
-    .written = static_cast<int>(result.written),
-  };
-}
-}  // namespace
-
 jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+  jsg::JsString str = input.orDefault(js.str());
+
+  if (str.length(js) == 0) {
+    return jsg::BufferSource(js, jsg::BackingStore::alloc<v8::Uint8Array>(js, 0));
+  }
+
+  // Allocate the output buffer and perform the conversion while ValueView is alive, but defer
+  // creating the V8 BufferSource until after ValueView is destroyed. This approach uses
+  // BackingStore::wrap with a custom disposer to avoid the copy overhead that would occur with
+  // BackingStore::from in the v8 sandbox, since from() copies data when it's not already in the
+  // sandbox. By using new/delete with wrap(), we maintain ownership semantics compatible with V8's
+  // C-style BackingStore API while avoiding the extra allocation and copy.
+  jsg::BackingStore backing = [&]() {
+    v8::String::ValueView value_view(js.v8Isolate, str);
+    size_t length = static_cast<size_t>(value_view.length());
+
+    if (value_view.is_one_byte()) {
+      // Fast path for Latin-1 encoded strings. V8 uses Latin-1 (ISO-8859-1) encoding internally
+      // for strings that contain only code points <= U+00FF. We need to convert to UTF-8.
+      auto data = reinterpret_cast<const char*>(value_view.data8());
+      size_t utf8_length = simdutf::utf8_length_from_latin1(data, length);
+      auto* output = new kj::Array<kj::byte>(kj::heapArray<kj::byte>(utf8_length));
+      [[maybe_unused]] auto written =
+          simdutf::convert_latin1_to_utf8(data, length, output->asChars().begin());
+      KJ_DASSERT(written == output->size());
+      return jsg::BackingStore::wrap<v8::Uint8Array>(output->begin(), output->size(),
+          [](void*, size_t, void* ptr) { delete reinterpret_cast<kj::Array<kj::byte>*>(ptr); },
+          output);
```

Comment:
This also doesn't work with the v8 sandbox. The memory for the BackingStore *has* to be owned by the sandbox, that's why `BackingStore::from` copies when the sandbox is used.